typing/typecore.ml

```diff
@@ -5941,12 +5941,19 @@ and type_expect_
       if Types.is_mutable label.lbl_mut then
         fatal_error
           "Typecore.type_expect_: unboxed record labels are never mutable";
+      let record_sort =
+        let jkinds =
+          Array.map (fun lbl -> lbl.lbl_jkind) label.lbl_all |> Array.to_list
+        in
+        let record_jkind = Jkind.Builtin.product ~why:Unboxed_record jkinds in
+        Jkind.sort_of_jkind record_jkind
+      in
```

Comment:
I think we should get the sort from the type of the actual record value here, rather than computing a new sort from the kinds in the declaration.  In the future we'll allow record declarations like:

```ocaml
type ('a : any) t = #{ x : int; y : 'a }
```

Here, we don't actually know the sort of the field `y`, but we can know the sort of the fields of specific instantiations of this type.

There are a couple ways you could get the sort from the record.  You can call `type_sort` on the type in `record` here (using `~fixed:false` - this is exactly the kind of place we'll need to constrain type variables to be representable when we allow such record declarations).  Or you could change `type_label_access` to return the sort, possibly by changing it to use `type_expect` on the record rather than `type_exp`, and making the expected type a new type variable whose kind is a sort variable (using such a type as the expected type is essentially just checking the result is representable).

Reply:
Thanks for the explanation, updated in 7ed17a022d34ae4d60f1a7205aba0f7c329b950.